zerver/lib/test_classes.py

```diff
@@ -1197,58 +1196,56 @@ def email_envelope_from(self, email_message: EmailMessage) -> str:
-        )
-        user_profile.date_joined = timezone_now() - timedelta(
-            days=(user_profile.realm.waiting_period_threshold - 1)
+        realm = get_realm("zulip")
+        admin_email = "iago"
+        moderator_email = "shiva"
+        member_email = "hamlet"
+        new_member_email = "othello"
+        guest_email = "polonius"
+
+        new_member_user = self.example_user(new_member_email)
+        do_set_realm_property(realm, "waiting_period_threshold", 1000, acting_user=None)
+        new_member_user.date_joined = timezone_now() - timedelta(
+            days=(realm.waiting_period_threshold - 1)
         )
-        self.assertFalse(validation_func())
-
-        # Ensure that the new moderators can also create streams because moderator
-        # being above the full member in role hierarchy.
-        do_change_user_role(user_profile, UserProfile.ROLE_MODERATOR, acting_user=None)
-        self.assertTrue(validation_func())
+        new_member_user.save()
 
-        do_change_user_role(user_profile, UserProfile.ROLE_MEMBER, acting_user=None)
-        user_profile.date_joined = timezone_now() - timedelta(
-            days=(user_profile.realm.waiting_period_threshold + 1)
+        member_user = self.example_user(member_email)
+        member_user.date_joined = timezone_now() - timedelta(
+            days=(realm.waiting_period_threshold + 1)
         )
-        self.assertTrue(validation_func())
+        member_user.save()
+
+        do_set_realm_property(realm, policy, Realm.POLICY_ADMINS_ONLY, acting_user=None)
+        self.assertTrue(validation_func(admin_email))
+        self.assertFalse(validation_func(moderator_email))
+        self.assertFalse(validation_func(member_email))
+        self.assertFalse(validation_func(new_member_email))
+        self.assertFalse(validation_func(guest_email))
+
+        do_set_realm_property(realm, policy, Realm.POLICY_MODERATORS_ONLY, acting_user=None)
+        self.assertTrue(validation_func(admin_email))
+        self.assertTrue(validation_func(moderator_email))
+        self.assertFalse(validation_func(member_email))
+        self.assertFalse(validation_func(new_member_email))
+        self.assertFalse(validation_func(guest_email))
+
+        do_set_realm_property(realm, policy, Realm.POLICY_MEMBERS_ONLY, acting_user=None)
+        self.assertTrue(validation_func(admin_email))
+        self.assertTrue(validation_func(moderator_email))
+        self.assertTrue(validation_func(member_email))
+        self.assertTrue(validation_func(new_member_email))
+        self.assertFalse(validation_func(guest_email))
+
+        do_set_realm_property(realm, policy, Realm.POLICY_FULL_MEMBERS_ONLY, acting_user=None)
+        self.assertTrue(validation_func(admin_email))
+        self.assertTrue(validation_func(moderator_email))
+        self.assertTrue(validation_func(member_email))
+        self.assertFalse(validation_func(new_member_email))
+        self.assertFalse(validation_func(guest_email))
```

Comment:
Let's move this block above the previous one so they are sorted in order of how wide a set of users get to do it.